### PR TITLE
Feature/custom user regex

### PR DESCRIPTION
**Pull request by MotwaniM**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -8,7 +8,7 @@ RESOURCE_PREFIX=rapid
 DOMAIN_NAME=example.com
 COGNITO_USER_POOL_ID=11111111
 LAYERS=raw,layer
-
+CUSTOM_USERNAME_REGEX="regex_expression"
```
Comment (TobyDrane) [inline .env.example:11]: Think we want to rename to `CUSTOM_USER_NAME_REGEX`

File: api/api/common/config/auth.py [hunk 2]
```diff
@@ -18,6 +18,8 @@
 COGNITO_RESOURCE_SERVER_ID = f"https://{DOMAIN_NAME}"
 COGNITO_USER_POOL_ID = os.environ["COGNITO_USER_POOL_ID"]
 
+CUSTOM_USERNAME_REGEX = os.environ.get("CUSTOM_USERNAME_REGEX")
```
Comment (TobyDrane) [inline api/api/common/config/auth.py:21]: See comment above

Comment (TobyDrane) [inline api/api/common/config/auth.py:21]: Don't think this wants to sit in the auth file, instead probably wants to move to the constants file

File: api/api/domain/user.py [hunk 3]
```diff
@@ -19,8 +27,15 @@ def get_validated_username(self):
         https://docs.aws.amazon.com/cognito/latest/developerguide/limits.html
         """
         if self.username is not None and re.fullmatch(USERNAME_REGEX, self.username):
-            return self.username
-        raise UserError("Invalid username provided")
+            custom_username_regex = os.environ.get("CUSTOM_USERNAME_REGEX")
+            if re.fullmatch(custom_username_regex, self.username):
```
Comment (TobyDrane) [inline api/api/domain/user.py:31]: What happens if they don't specify a custom user name regex? We only want to perform a check if they have specified a custom regex

File: api/test/api/adapter/test_cognito_adapter.py [hunk 4]
```diff
@@ -247,6 +248,9 @@ def test_throws_error_when_client_app_name_has_not_been_changed_from_placeholder
 
 
 class TestCognitoAdapterUsers(BaseCognitoAdapter):
+    @mock.patch.dict(
+        os.environ, {"CUSTOM_USERNAME_REGEX": "[a-zA-Z][a-zA-Z0-9@._-]{2,127}"}
```
Comment (TobyDrane) [inline api/test/api/adapter/test_cognito_adapter.py:252]: I don't think we want to patch the variable here but instead enforce it by ensuring it's set correctly within the environment variables when running the tests.

We will want a test where we patch it to be None and ensure we test the counter-factual of when they don't set it and thus we just fallback to the default validation

File: infrastructure/modules/app-cluster/variables.tf [hunk 5]
```diff
@@ -202,3 +202,17 @@ variable "layers" {
   description = "A list of the layers that the rAPId instance will contain"
   default     = ["default"]
 }
+
+variable "custom_username_regex" {
+  type        = list(string)
+  description = "A list containing the regex expression for conditional user validation and a string to test against. The input must be a terraform list variable composed of two strings. The first string is the regex expression and the second string is the test string for said regular expression."
+  # Set default to a regex expression for basic api username validation
+  default = ["[a-zA-Z][a-zA-Z0-9@._-]{2,127}", "AUser1"]
```
Comment (TobyDrane) [inline infrastructure/modules/app-cluster/variables.tf:210]: We want a default to be None and this to become a optional variable. Also doesn't make sense for this to be a list of strings as we just want this to be a regex string

File: infrastructure/modules/app-cluster/main.tf [hunk 6]
```diff
@@ -9,6 +9,7 @@ locals {
     "COGNITO_USER_POOL_ID" : var.cognito_user_pool_id,
     "RESOURCE_PREFIX" : var.resource-name-prefix,
     "COGNITO_USER_LOGIN_APP_CREDENTIALS_SECRETS_NAME" : var.cognito_user_login_app_credentials_secrets_name
+    "CUSTOM_USERNAME_REGEX" : var.custom_username_regex
```
Comment (TobyDrane) [inline infrastructure/modules/app-cluster/main.tf:12]: Needs to handle the None type and ensure if we pass None here that Python correctly picks up the None

**Review by TobyDrane — COMMENTED** [review]
@MotwaniM reviewed, shout if it doesn't make sense. Also we'll need documentation for this writing - https://rapid.readthedocs.io/en/latest/

File: infrastructure/modules/app-cluster/main.tf (hunk 6 above)
Comment (MotwaniM) [reply infrastructure/modules/app-cluster/main.tf]: Have made a fix. Uses Null values which should omit the argument from the container and force a python None value when using os.environ.get

File: .github/workflows/dev.yml [hunk 7]
```diff
@@ -45,9 +45,9 @@ jobs:
       - name: Populate .env with additional vars
         run: |
           cp ./.github/.github.env .env
-          echo AWS_ACCOUNT=${{ secrets.AWS_ACCOUNT }} >> .env
-          echo AWS_REGION=${{ secrets.AWS_REGION }} >> .env
-          echo AWS_DEFAULT_REGION=${{ secrets.AWS_REGION }} >> .env
+          echo \AWS_ACCOUNT=${{ secrets.AWS_ACCOUNT }} >> .env
```
Comment (TobyDrane) [inline .github/workflows/dev.yml:48]: Do we need these backslashes?

File: api/api/common/config/constants.py [hunk 8]
```diff
@@ -39,3 +40,8 @@
 
 FIRST_SCHEMA_VERSION_NUMBER = 1
 SCHEMA_VERSION_INCREMENT = 1
+
+CUSTOM_USER_NAME_REGEX_ENVIRON = os.environ.get("CUSTOM_USER_NAME_REGEX")
```
Comment (TobyDrane) [inline api/api/common/config/constants.py:44]: For this we can use the default Python getenv that returns None if the environment variable doesn't exist by default

https://docs.python.org/3/library/os.html#os.getenv

File: docs/infrastructure/deployment.md [hunk 9]
```diff
@@ -56,7 +56,7 @@ There are also these optional inputs:
 - `catalog_disabled` - if set to `true` it will disable the rAPId internal data catalogue
 - `tags` - if provided, it will tag the resources with the defined value. Otherwise, it will default to "Resource = '
   data-f1-rapid'"
-
+- `custom_user_name_regex` - Regex to allow custom filtering of usernames supplied when creating a new user. Defaults to null, in which case rAPId will default to its basic username validity checks.
```
Comment (TobyDrane) [inline docs/infrastructure/deployment.md:59]: "Regex that when supplied usernames must conform to when creating a new user. Defaults to none, in which case rAPId will default to it's basic username validity checks."

**Review by TobyDrane — CHANGES_REQUESTED** [review]

File: .github/workflows/dev.yml (hunk 7 above)
Comment (MotwaniM) [reply .github/workflows/dev.yml:48]: Yeah, there's a formatting issue with the .env file where AWS_ACCOUNT was ending up on the end of the last line of the env file; this was causing the regex tests to fail.

File: api/api/common/config/constants.py (hunk 8 above)
Comment (MotwaniM) [reply api/api/common/config/constants.py:44]: Apologies, this code was a holdover from when I was fixing the actions. Removed and changed to .getenv

File: .github/workflows/dev.yml (hunk 7 above)
Comment (TobyDrane) [reply .github/workflows/dev.yml:48]: Hmm not entirely sure how/why this has now suddenly cropped in as an issue? Think we probably want to get the bottom of this as not ideal having to add this backslash in, might be we have to wrap our env variables with quotes or similar?

**Pull request merged**